src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Stream.cs

```diff
@@ -45,76 +42,96 @@ private static async ValueTask<TValue> ReadAsync<TValue>(Stream utf8Json, Type r
+                    {
+                        int bytesRead = await utf8Json.ReadAsync(buffer.AsMemory(bytesInBuffer), cancellationToken).ConfigureAwait(false);
+
+                        if (bytesRead == 0)
+                        {
+                            isFinalBlock = true;
+                            break;
+                        }
 
-                    deserializeBufferSize = bytesRemaining + bytesRead;
-                    isFinalBlock = (bytesRead == 0);
+                        totalBytesRead += bytesRead;
+                        bytesInBuffer += bytesRead;
 
+                        if (bytesInBuffer > clearMax)
+                        {
+                            clearMax = bytesRead;
+                        }
+
+                        if (bytesInBuffer == buffer.Length)
+                        {
+                            break;
+                        }
+                    }
+
+                    // Process the data available
                     ReadCore(
                         ref readerState,
                         isFinalBlock,
-                        buffer,
-                        deserializeBufferSize,
+                        new Span<byte>(buffer, 0, bytesInBuffer),
                         options,
                         ref state);
 
                     int bytesConsumed = (int)readerState.BytesConsumed;
-                    bytesRemaining = deserializeBufferSize - bytesConsumed;
+                    bytesInBuffer -= bytesConsumed;
 
                     if (isFinalBlock)
                     {
                         break;
                     }
 
                     // Check if we need to shift or expand the buffer because there wasn't enough data to complete deserialization.
-                    if (bytesConsumed <= (bufferSize / 2))
+                    if ((uint)bytesInBuffer > ((uint)buffer.Length / 2))
                     {
                         // We have less than half the buffer available, double the buffer size.
-                        bufferSize = (bufferSize < HalfMaxValue) ? bufferSize * 2 : int.MaxValue;
+                        byte[] dest = ArrayPool<byte>.Shared.Rent((buffer.Length < (int.MaxValue / 2)) ? buffer.Length * 2 : int.MaxValue);
+                        
+                        // Copy the unprocessed data to the new buffer while shifting the processed bytes.
+                        Buffer.BlockCopy(buffer, bytesConsumed, dest, 0, bytesInBuffer);
 
-                        byte[] dest = ArrayPool<byte>.Shared.Rent(bufferSize);
-                        bufferSize = dest.Length;
-                        if (bytesRemaining > 0)
-                        {
-                            // Copy the unprocessed data to the new buffer while shifting the processed bytes.
-                            Buffer.BlockCopy(buffer, bytesConsumed, dest, 0, bytesRemaining);
-                        }
+                        new Span<byte>(buffer, 0, clearMax).Clear();
```

Comment:
nit: `buffer.AsSpan(0, clearMax).Clear();`

Same elsewhere.

Reply:
Is that any faster?  It seems like it's functionally exactly the same but put extra pressure on the JIT to inline.